src/dune_rules/mdx.ml

```diff
@@ -229,17 +271,98 @@ let gen_rules_for_single_file stanza ~sctx ~dir ~expander ~mdx_prog src =
+  let dune_version = Scope.project scope |> Dune_project.dune_version in
+  let file = Path.Build.relative dir "mdx_gen.ml-gen" in
+
+  (* Libs from the libraries field should have their include directories sent to
+     mdx *)
+  let open Resolve.O in
+  let directory_args =
+    let+ libs_to_include =
+      Resolve.List.filter_map t.libraries ~f:(function
+        | Direct lib
+        | Re_export lib ->
+          let+ lib = Lib.DB.resolve (Scope.libs scope) lib in
+          Some lib
+        | _ -> Resolve.return None)
+    in
+
+    let mode = Context.best_mode (Super_context.context sctx) in
+
+    let libs_include_paths = Lib.L.include_paths libs_to_include mode in
+
+    let open Command.Args in
+    let args =
+      Path.Set.to_list libs_include_paths
+      |> List.map ~f:(fun p -> S [ A "--directory"; Path p ])
+    in
+    S args
+  in
+
+  let prelude_args =
+    Command.Args.S (List.concat_map t.preludes ~f:(Prelude.to_args ~dir))
+  in
+
+  (* We call mdx to generate the testing executable source *)
+  let action =
+    Command.run ~dir:(Path.build dir) mdx_prog ~stdout_to:file
+      [ A "dune-gen"; prelude_args; Resolve.args directory_args ]
+  in
+  let open Memo.Build.O in
+  let* () = Super_context.add_rule sctx ~loc ~dir action in
+
+  (* We build the generated executable linking in the libs from the libraries
+     field *)
+  let obj_dir = Obj_dir.make_exe ~dir ~name in
+  let main_module_name = Module_name.of_string name in
+  let module_ = Module.generated ~src_dir:(Path.build dir) main_module_name in
+  let modules = Modules.singleton_exe module_ in
+  let flags = Ocaml_flags.default ~dune_version ~profile:Release in
+  let compile_info =
+    Lib.DB.resolve_user_written_deps_for_exes (Scope.libs scope)
+      [ (t.loc, name) ]
+      (Lib_dep.Direct (loc, Lib_name.of_string "mdx.test") :: t.libraries)
+      ~pps:[ (*Preprocess.Per_module.pps t.preprocess*) ] ~dune_version
+  in
+  let cctx =
+    Compilation_context.create ~super_context:sctx ~scope ~expander ~obj_dir
+      ~modules ~flags
+      ~requires_compile:(Lib.Compile.direct_requires compile_info)
+      ~requires_link:(Lib.Compile.requires_link compile_info)
+      ~opaque:(Explicit false) ~js_of_ocaml:None ~package:None ()
+  in
+  let+ () =
+    Exe.build_and_link cctx
+      ~program:{ name; main_module_name; loc }
+      ~link_args:(Action_builder.return (Command.Args.A "-linkall"))
+      ~linkages:[ Exe.Linkage.byte ] ~promote:None
+  in
+  Path.Build.relative dir (name ^ ".bc")
+
 (** Generates the rules for a given mdx stanza *)
-let gen_rules t ~sctx ~dir ~expander =
+let gen_rules t ~sctx ~dir ~scope ~expander =
   let open Memo.Build.O in
   let register_rules () =
     let* files_to_mdx = files_to_mdx t ~sctx ~dir
     and* mdx_prog =
       Super_context.resolve_program sctx ~dir ~loc:(Some t.loc)
         ~hint:"opam install mdx" "ocaml-mdx"
     in
+    let* mdx_prog_gen =
+      if Dune_lang.Syntax.Version.Infix.(t.version >= (0, 2)) then
```

Comment:
Is there a way to use `Memo.Build.if_` for something non-unit?

Reply:
@jeremiedimino  ?

Reply:
Not with the current `if_` function. It could be generalised as follow:

```ocaml
val if_ : bool t -> then_:(unit -> 'a t) -> else_:(unit -> 'a t) -> 'a t

let if_ cond ~then_ ~else_ =
  cond >>= function
  | true -> then_
  | false -> else_
```

I personally find the expansion clear enough and wouldn't introduce such an `if_` function.

It seems to me that the current `if_` should be renamed `when_`. @snowleopard what do you think?

Reply:
Ah, I suggested that because I assumed that `if_` is more efficient than binding into an OCaml `if` (because it gives a static approximation of the result). Is that not the case?

Reply:
We used to do things like this, but not anymore. It was putting too much pressure on the code.

The only rule that still exist is: you might get more parallelism if you write `let* ... and* ... in` rather than `let* ... in let* ... in`. But sometimes we still use the second version to make the code clearer.

Reply:
> It seems to me that the current `if_` should be renamed `when_`. @snowleopard what do you think?

@jeremiedimino Yes, `when_` is better name for the combinator `bool t -> (() -> unit t) -> unit t`. Another possible name for it is `Bool.iter`. The signature is similar to `Option.iter`: `bool` is isomorphic to `unit option`, so `Bool.iter` would run the callback if the contained value is `true` :-) But `when_` is probably better.

Reply:
Makes sense. I personally prefer `when_`. It seems simpler and clear enough to me, and I have seen it in at least one other language.

Reply:
Renamed to `when_` in #4766.